### DISCUSSION TITLE
50万数据平均耗时正常范围200-300ms，但是有部分请求会突然变成10s，原因是啥

### DISCUSSION THREAD

**Discussion by jhyever** [milvus-io/milvus#38588] [Announcement]
<img width="274" alt="0eb9ef071aef7e7b3b8afffd5d01492" src="https://github.com/user-attachments/assets/c2073e96-1167-4b1d-90e8-4f14d1dcd7b4" />

**Comment by yhmo** [comment]
多半是客户端并发量达到了milvus server处理能力的上限，有些请求会在proxy那里排队等很久。

**Reply by jhyever** [reply]
这个有对应的避免办法麽，会有请求时间一下子变长，又回归正常

**Reply by jhyever** [reply]
有解决方案麽

**Reply by yhmo** [reply]
配置更好的CPU，或者查询参数改低点(比如用了IVF索引的话，把nprobe参数改低点，会降低召回率)
或者改别的索引试试，HNSW会比较快一点

打个比方，家用车不太可能跑到300公里时速，油门踩到底也不行，你觉得有没有方法让普通轿车开到300km/h以上？

**Reply by jhyever** [reply]
但是平均都是在200毫秒左右的，突然个别就10秒了，感觉就有问题，并发路数也就是10路、20路呢，麻烦帮忙解答一下，急，谢谢

**Reply by jhyever** [reply]
是不是哪里有并发的路数限制或是配置 ，单机版的v2.4.6版本

**Reply by yhmo** [reply]
应该不是个别10秒。你把区间都统计一下看看，可能9秒8秒的也有。可能某个时间点进去的几十几百个请求都有很长的latency。

你可以把milvus.yaml里的queryNode.grouping.enabled改成false再试试
```
queryNode:
  grouping:
    enabled: false
    maxNQ: 1000
    topKMergeRatio: 20
```

**Reply by jhyever** [reply]
我们集成了这个要坐性能压测测试，
第一次50万数据，10万请求，10路并发，有23条请求在5-10秒区间
第二次50万数据，6万请求，10路并发，有17条在5-10秒区间
就是这种情况感觉不正常，但是平均都是在200毫秒左右

1、单机是standalone部署的，没有values.yaml文件，怎么设置queryNode.grouping.enabled呀  @yhmo 

补充一点：代码的sdk是：
            <groupId>io.milvus</groupId>
            <artifactId>milvus-sdk-java</artifactId>
            <version>2.3.4</version>

部署的milvus版本是2.4.6

2、milvus server处理能力的上限是什么呢

**Reply by yhmo** [reply]
standalone的配置: https://milvus.io/docs/configure-docker.md?tab=component

**Reply by jhyever** [reply]
这个问题呢
1、第一次50万数据，10万请求，10路并发，有23条请求在5-10秒区间
第二次50万数据，6万请求，10路并发，有17条在5-10秒区间
就是这种情况感觉不正常，但是平均都是在200毫秒左右    就是一些个别的

2、单机版的milvus server处理能力的上限是什么呢

**Reply by yhmo** [reply]
1. 我比较倾向于是达到了查询能力的上限，随着并行压力越大，长延时的请求数量会越来越多，最长延时会越来越长。
2. 查询的性能跟很多因素相关，cpu，内存，网络带宽，数据量，向量维度，过滤条件，topk，output_fields，索引类型，索引参数，请求压力，等等。具体在某个条件下的上限只能实测得出，没人能估计出具体的值。

**Reply by jhyever** [reply]
milvus的查询默认处理能力上限是多少呢，上面那个性能表格，就是调用milvus检索开始和拿到响应时的时间耗时

**Reply by yhmo** [reply]
100万数据以下单个查询请求耗时一般是在10ms以内。
但是，如果你是多并发，事情就不一样了。查询是会合并执行的，就好比公交车站来一辆车运走十几个乘客一样，公交车会比的士慢，但载人效率高。

**Reply by jhyever** [reply]
但是我们平均都是在200-300毫秒左右，部分会在10秒左右，
我们用的索引算法是FLAT算法，sdk用的是2.3.5     部署的版本用的是2.4.6，
跟索引算法有关么，

**Reply by yhmo** [reply]
其实就是公交车和的士的区别，你乘客多了，公交车上下车耗时多，开得慢，每个乘客等的时间拉长。
如果只有几个乘客，一辆的士直接开到目的地，肯定快很多。

证明的方法很多，第一，你可以不用多个client，只单独发一次search请求，看是不是10ms左右。
第二，你可以禁止合并查询，规定每个请求都单独执行，在milvus.yaml里这么设置，然后只用一个客户端发查询请求，看是不是10ms
```
queryNode:
  grouping:
    enabled: false
    maxNQ: 1000
    topKMergeRatio: 20
```

**Reply by yhmo** [reply]
FLAT是暴搜，每一条向量都会被计算，单个请求耗时50ms也是正常的，50万条向量
换成IVF_FLAT，nlist设置128，nprobe设置8，耗时就10ms以下。或者换成HNSW

**Reply by jhyever** [reply]
合并查询 是啥意思呀，会影响功能嘛
queryNode:
  grouping:
    enabled: false
    maxNQ: 1000
    topKMergeRatio: 20

**Reply by yhmo** [reply]
合并查询就是车站等车的乘客太多，如果用的士一个个运，又费钱又费油，所以换成大巴运，一辆大巴50人运走，省钱。
https://zhuanlan.zhihu.com/p/148274838 这是早期milvus的合并查询机制，跟现在的milvus的合并查询原理也差不多

**Reply by jhyever** [reply]
这样话，合并查询性能应该更好的才对呀，我看我配置现在就是合并查询，如果改成false，岂不是性能又变差了嘛，
现在主要就是个别请求会耗时较长

**Reply by yhmo** [reply]
性能有两个方面： 
1. 每个查询的耗时越短越好
2. 每秒能执行的查询请求数量越多越好

合并查询会使得每个查询的耗时延长，但每秒执行的请求数量增加了，那你说这是性能变好还是变差？
